network: introduce an online state that respects RequiredForOnline=

sipraga opened the pull request:
resolves #19143

This PR adds a new "online" state to systemd-networkd which indicates whether a link is online based on the `RequiredForOnline=` property. Moreover, it exposes an overall "online" state through the networkd manager which is a survey the online state of all "required for online" links. Finally, it updates the `network_is_online()` utility function to make use of this state if it is meaningful. More detail can be found in the commit messages.

The motivation behind this change is to give a more meaningful online status to timesyncd - the only user of `network_is_online()`. An administrator can then configure `RequiredForOnline=no` on links which do not provide any gateway access, which prevents timesyncd from attempting spurious timesyncs when such links enter a degraded or routable state. This in turn makes for a faster timesync when a suitable link later comes online, as we have not exhausted timesyncd's internal rate limiter meanwhile. A detailed description of the problem is in #19143.

poettering commented on src/network/networkd-state-file.c:
```diff
@@ -136,6 +137,20 @@ int manager_save(Manager *m) {
+                        else if (online_state < link->online_state) {
+                                /* push up the online state */
+                                online_state = link->online_state;
+                        }
```
coding style, plesae place `else if` on same line as closing `}`. And no {} around single line if blocks.

https://systemd.io/CODING_STYLE/#formatting

sipraga replied:
This original implementation was hard to read and had a bug, so I went for an alternative solution. Please check.

poettering commented on src/libsystemd/sd-network/network-util.h:
```diff
@@ -44,6 +44,15 @@ typedef enum LinkAddressState {
+        LINK_ONLINE_STATE_PARTIALLY_ONLINE,
+        LINK_ONLINE_STATE_ONLINE,
+        _LINK_ONLINE_STATE_MAX,
+        _LINK_ONLINE_STATE_INVALID = -EINVAL,
```
hmm, do we really want both LINK_ONLINE_STATE_UNKNOWN and _LINK_ONLINE_STATE_INVALID? The latter should suffice as marker for "dont know", no?

sipraga replied:
the intent here was to make it a "first class citizen" among statuses - it seems a little weird to repurpose _LINK_ONLINE_STATE_INVALID for that. I'll try to change it in the next revision and we can see how it looks

sipraga replied:
I removed the unknown state and use _LINK_ONLINE_STATE_INVALID now. Seems to work just fine.

poettering reviewed (changes requested):
some comments. I figure @yuwata should have a look at this and decide

sipraga commented on src/libsystemd/sd-network/network-util.c:
```diff
@@ -64,6 +64,15 @@ static const char* const link_address_state_table[_LINK_ADDRESS_STATE_MAX] = {
+static const char *const link_online_state_table[_LINK_ONLINE_STATE_MAX] = {
+        [LINK_ONLINE_STATE_UNKNOWN]          = "unknown",
+        [LINK_ONLINE_STATE_OFFLINE]          = "offline",
+        [LINK_ONLINE_STATE_PARTIALLY_ONLINE] = "partially-online",
```
"partial" may also be preferable to "partially"?

sipraga replied:
I went for "partial" in the end

sipraga commented:
Changes:
- remove `LINK_ONLINE_STATE_UNKNOWN` and use `_LINK_ONLINE_STATE_INVALID` instead
- only compute online state for managed links with `RequiredForOnline=yes`
- fix a bug in the calculation of manager online state, and in general make it easier to read
- update documentation ever so slightly
- update networkctl to display online state for links/manager

The only significant change here is that I don't bother to compute an online state for links where `RequiredForOnline=no`. I think this is fair enough because for such links, we would fall back to the `LINK_OPERSTATE_RANGE_DEFAULT` range anyway, which could be potentially misleading. If a user sets `RequiredForOnline=no`, then the online state is useless anyway.

I think this fine for another review.

sipraga commented:
Something else to think about: Perhaps there is room for a networkd.conf `[Link]` option `AnyForOnline=` (or similar), which will set the manager online state to `online` in case any required link is online. A bit like the `--any` switch for systemd-networkd-wait-online. If false, then the state would be `partial` as is the case at the moment.

I haven't touched the wait-online code here, but with such a `AnyForOnline=` option for networkd, maybe most of the extra switches to wait-online program can be removed and the service simplified to just wait on the overall manager online state. At the moment it does a lot of careful checking but I do wonder what the advantage of this is, when the administrator could just configure the same ranges and preferences in the .network files?

Anyway, those are just some personal thoughts of mine. The current PR doesn't affect wait-online in any way, nor was that the original aim.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -242,6 +243,15 @@ void link_update_operstate(Link *link, bool also_update_master) {
 
+        /* Only determine online state for managed links with RequiredForOnline=yes */
+        if (!link->network || !link->network->required_for_online)
+                online_state = _LINK_ONLINE_STATE_INVALID;
```
Hmm, I think we should also set online state for unmanaged links.
How about the following?
```
if (operstate >= link->network ? link->network->required_operstate_for_online.min : LINK_OPERSTATE_RANGE_DEFAULT.min &&
    operstate <= link->network ? link->network->required_operstate_for_online.max : LINK_OPERSTATE_RANGE_DEFAULT.max)
        online_state = LINK_ONLINE_STATE_ONLINE;
else
        online_state = LINK_ONLINE_STATE_OFFLINE;
```

sipraga replied:
Hmm, what value would that bring? For unmanaged links, `LINK_OPERSTATE_RANGE_DEFAULT` is only a best guess on the part of networkd, since it does not know how the link should be configured. But it can be potentially misleading too. Why not be transparent about it?

For managed links where `RequiredForOffline=no`, I would argue the same way. By setting `=no`, the admin has no way to configure the actual `required_operstate_for_online`, so it will be a best guess.

I can of course do what you suggest - I don't think it's wrong per se - but I invite you to reconsider first.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -2337,6 +2354,7 @@ static int link_initialized_and_synced(Link *link) {
                 }
 
                 link->network = network_ref(network);
+                link_update_operstate(link, false);
```
I do not think this is necessary.

sipraga replied:
I put this here for managed links which have no carrier at boot. `link_update_operstate()` needs to get called at least once after the `link->network` pointer has been set, so that the online state can transition from `_LINK_ONLINE_STATE_INVALID` to `LINK_ONLINE_STATE_OFFLINE`.

Put another way, we want managed links which are offline to be marked as such. Without this line, such links which never have carrier may be marked as "n/a" (or "unknown", etc.).

I'm not sure I see an alternative, except to have the default online state of a link be `LINK_ONLINE_STATE_OFFLINE`. But I am trying to make networkd as transparent as possible when it advertises the online state. Hence I prefer to say "networkd has no idea", than "networkd thinks this is offline".

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -105,10 +105,12 @@ static int ordered_set_put_in4_addrv(
         LinkOperationalState operstate = LINK_OPERSTATE_OFF;
         LinkCarrierState carrier_state = LINK_CARRIER_STATE_OFF;
         LinkAddressState address_state = LINK_ADDRESS_STATE_OFF;
+        LinkOnlineState online_state = _LINK_ONLINE_STATE_INVALID;
```
Please drop the initialization. Also see the comment below.

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -217,6 +226,13 @@ int manager_save(Manager *m) {
         if (carrier_state >= LINK_CARRIER_STATE_ENSLAVED)
                 carrier_state = LINK_CARRIER_STATE_CARRIER;
 
+        if (links_online && !links_offline)
+                online_state = LINK_ONLINE_STATE_ONLINE;
+        else if (links_online && links_offline)
+                online_state = LINK_ONLINE_STATE_PARTIAL;
+        else if (links_offline)
+                online_state = LINK_ONLINE_STATE_OFFLINE;
```
First, `link_online` and `links_offline` are integer, do not implicitly cast to boolean.

How about the following?
```
online_state = links_online > 0 ? (links_offline > 0 ? LINK_ONLINE_STATE_PARTIAL : LINK_ONLINE_STATE_ONLINE)
                                : (links_offline > 0 ? LINK_ONLINE_STATE_OFFLINE : _LINK_ONLINE_STATE_INVALID);
```

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -419,9 +445,12 @@ int link_save(Link *link) {
                 bool space;
 
+                online_state = link_online_state_to_string(link->online_state);
+                fprintf(f, "ONLINE_STATE=%s\n", strempty(online_state));
```
I think it is better to not show `ONLINE_STATE=` if the string is empty:
```
if (online_state)
        fprintf(f, "ONLINE_STATE=%s\n", online_state);
```

yuwata commented on src/libsystemd/sd-network/network-util.c:
```diff
@@ -13,20 +13,32 @@
 
 bool network_is_online(void) {
-        _cleanup_free_ char *carrier_state = NULL, *addr_state = NULL;
+        _cleanup_free_ char *online_state = NULL, *carrier_state = NULL, *addr_state = NULL;
```
Please move the declarations of `carrier_state` and `addr_state` to the `else if (state < 0)` block below.

yuwata commented on src/network/networkctl.c:
```diff
@@ -712,6 +730,8 @@ static int list_links(int argc, char *argv[], void *userdata) {
                                    TABLE_STRING, strna(t),
                                    TABLE_STRING, strna(operational_state),
                                    TABLE_SET_COLOR, on_color_operational,
+                                   TABLE_STRING, strna(online_state),
```
Maybe, instead of `n/a`, `unknown` is better?
```
TABLE_STRING, online_state ?: "unknown",
```

yuwata commented on src/network/networkctl.c:
```diff
@@ -1483,6 +1506,13 @@ static int link_status_one(
+        r = table_add_many(table,
+                           TABLE_EMPTY,
+                           TABLE_STRING, "Online state:",
+                           TABLE_STRING, strna(online_state),
```
Same here.

yuwata commented on src/network/networkctl.c:
```diff
@@ -2176,7 +2209,11 @@ static int system_status(sd_netlink *rtnl, sd_hwdb *hwdb) {
+                           TABLE_SET_COLOR, on_color_operational,
+                           TABLE_EMPTY,
+                           TABLE_STRING, "Online state:",
+                           TABLE_STRING, strna(online_state),
```
Also here.

yuwata commented on src/network/networkd-link.h:
```diff
@@ -75,6 +75,7 @@ typedef struct Link {
         LinkOperationalState operstate;
         LinkCarrierState carrier_state;
         LinkAddressState address_state;
+        LinkOnlineState online_state;
```
I think, it is better to initialize this with `_LINK_ONLINE_STATE_INVALID` in `link_new()`.

yuwata commented on src/network/networkd-manager.h:
```diff
@@ -39,6 +39,7 @@ struct Manager {
         LinkOperationalState operational_state;
         LinkCarrierState carrier_state;
         LinkAddressState address_state;
+        LinkOnlineState online_state;
```
This should be also initialized with `_LINK_ONLINE_STATE_INVALID` in `manager_new()`.

yuwata reviewed (changes requested):
Sorry for late review.

yuwata commented on src/network/networkctl.c:
```diff
@@ -691,14 +705,18 @@ static int list_links(int argc, char *argv[], void *userdata) {
-                _cleanup_free_ char *setup_state = NULL, *operational_state = NULL;
+                _cleanup_free_ char *setup_state = NULL, *operational_state = NULL, *online_state = NULL;
                 const char *on_color_operational, *off_color_operational,
+                           *on_color_online, *off_color_online,
```
`off_color_online` is not used. See #19288.

sipraga replied:
I'll rebase and fix once #19288 is merged.

sipraga commented:
Hi @yuwata, thanks for your review. I have replied to two of your comments where I would like some extra feedback before I change the code. In the mean time I will address the other comments and push an updated change.

sipraga commented:
- rebased
- initialized `.online_state` of manager/link in `{manager,link}_new()`
- display `"unknown"` instead of `"n/a"` in networkctl
- don't print `ONLINE_STATE=` to static state file when empty
- change manager `online_state` derivation to not use ints as bools; set `_LINK_ONLINE_STATE_INVALID` explicitly

sipraga commented:
- added missing newline

yuwata commented:
Hmm, after reconsidering the above comments, now I am skeptical about introducing the per link online state. The link online state can be easily obtained from operstate and the required operstate rage of the link.

So, how about to introduce online state only on manager? Something like the following.
```
static int link_is_online(Link *link) {
        assert(link);

        if (!link->network || link->network->unmanaged || !link->network->required_for_online)
                return -ENODATA; /* arbitrary negative errno. */

        return operstate >= link->network->required_operstate_for_online.min &&
                operstate <= link->network->required_operstate_for_online.max;
}

int manager_save(Manager *m) {
        ...
        HASHMAP_FOREACH(link, m->link) {
                ...
                r = link_is_online(link);
                if (r > 0)
                        links_online++;
                else if (r == 0)
                        links_offline++;
}
```
Then, I think the logic seems quite simple.

sipraga commented:
> Hmm, after reconsidering the above comments, now I am skeptical about introducing the per link online state. The link online state can be easily obtained from operstate and the required operstate rage of the link.
> 
> So, how about to introduce online state only on manager? Something like the following.

As far as this PR is concerned, the advantage of a per-link online state is better visibility for the administrator. For example, a misconfigured system might have manager state "partial" when the expected state is "online", but without a per-link online state it can be a little tricky to identify the offline link(s) responsible for this. With the way I have implemented it right now, the administrator can just run `networkctl` and inspect all the "offline" links to narrow down the problem. This is another example of why I think it's good to distinguish between "unknown" and "offline" states for unmanaged/managed links.

Do you find that persuasive enough? If not then, I think your suggestion is also fine. But in that case, should I rename the state enum to `ManagerOnlineState`, `MANAGER_ONLINE_STATE_OFFLINE`, etc.?

sipraga commented:
Hi @yuwata. Considering my comment above, would you still prefer just a manager-only online state?

yuwata commented:
@sipraga Sorry for late response. Now I am re-re-considering the feature. I will comment later.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -242,6 +243,15 @@ void link_update_operstate(Link *link, bool also_update_master) {
+        /* Only determine online state for managed links with RequiredForOnline=yes */
+        if (!link->network || !link->network->required_for_online)
+                online_state = _LINK_ONLINE_STATE_INVALID;
+        else if (operstate >= link->network->required_operstate_for_online.min &&
+                 operstate <= link->network->required_operstate_for_online.max)
```
Could you also take into account `ipv4/6_address_state` and `required_address_family`?

yuwata reviewed (changes requested):
Sorry for late reply. After re-re-consideration, now I think your points make sense and this is quite useful. Please go ahead in the current form.

sipraga commented:
- rebased
- removed `off_color_online` variables, which were never used
- now take into account `ipv{4,6}_address_state` and `required_address_family` when determining link online status
- update documentation for `RequiredFamilyForOnline=` accordingly

sipraga commented:
- fixed missing `</para>` in man page xml

yuwata commented on src/network/networkd-link.c:
```diff
@@ -256,6 +257,38 @@ void link_update_operstate(Link *link, bool also_update_master) {
                 operstate = LINK_OPERSTATE_ENSLAVED;
 
+        /* Only determine online state for managed links with RequiredForOnline=yes */
+        if (!link->network || !link->network->required_for_online)
```
Please add one more condition about unmanaged.
```
if (!link->network || link->network->unmanaged || !link->network->required_for_online)
```

yuwata commented on src/network/networkctl.c:
```diff
@@ -120,6 +120,25 @@ static void setup_state_to_color(const char *state, const char **on, const char
         }
 }
 
+static void online_state_to_color(const char *state, const char **on, const char **off) {
```
Please update the commit message 'n/a' -> 'unknown'.

yuwata reviewed (changes requested):
Looks mostly good for me. Two minor comments.

sipraga commented:
- rebased
- added check for `link->network->unmanaged`
- updated networkctl commit message 'n/a' -> 'unknown'

yuwata commented:
LGTM, but all Ubuntu CIs fail...

sipraga commented:
Hmm, looks like:
- networkd-dispatcher.service fails for some reason (not familiar with this service)
- networkd-test.py fails on some regex on the state file

Sorry for this. I will investigate and ping you when it's fixed. Thanks for your patience!

sipraga commented:
@yuwata with the above patch, `networkd-test.py` passes fine. But the Ubuntu CIs are still failing due to `networkd-dispatcher.service` failing. I have tried to get the test environment running properly on my (Arch Linux) system, but to no avail. Any tips on what might be causing this service to fail?

mrc0mmand commented:
> @yuwata with the above patch, `networkd-test.py` passes fine. But the Ubuntu CIs are still failing due to `networkd-dispatcher.service` failing. I have tried to get the test environment running properly on my (Arch Linux) system, but to no avail. Any tips on what might be causing this service to fail?

I've opened https://salsa.debian.org/systemd-team/systemd/-/merge_requests/127 which, once merged, should make the `networkd-dispatcher.serve` fail more obvious.

yuwata commented:
networkd-dispatcher needs to be updated.
https://gitlab.com/craftyguy/networkd-dispatcher/-/issues/52

yuwata commented on src/network/networkctl.c:
```diff
@@ -684,7 +703,7 @@ static int list_links(int argc, char *argv[], void *userdata) {
         (void) pager_open(arg_pager_flags);
 
-        table = table_new("idx", "link", "type", "operational", "setup");
+        table = table_new("idx", "link", "type", "operational", "online", "setup");
```
Not sure, but maybe "online" state at the end maybe ok for networkd-dispatcher.service??

sipraga replied:
I did this, but it didn't seem to help... I'll leave it at the end of the line now though, since it doesn't really matter.

craftyguy commented:
Should this be documented in the networkctl manpage? 
https://github.com/systemd/systemd/blob/main/man/networkctl.xml#L131

sipraga commented:
Thanks all for the comments. I'll try putting the `ONLINE` column last and see we can get networkd-dispatcher to start up in that case - doesn't make a big difference I think.

I'll also push a commit for the networkctl doc.

sipraga commented:
- moved ONLINE column to the rightmost position
- added documentation to networkctl(1)

Let's see if this works now...

sipraga commented:
- fixed test

sipraga commented:
@yuwata all comments addressed, but it looks like networkd-dispatcher is still failing. I saw your issue but I also agree with @craftyguy that it doesn't necessarily make sense to extend that if the feature is not yet merged.

Do you think we should merge despite the Ubuntu CI failures?

craftyguy commented:
@sipraga if you have the ability to point your CI to a specific branch of networkd-dispatcher, I could make a patch and put it there (but not merge it to the master branch, per my earlier comment on that issue)

sipraga commented:
@craftyguy I had a look but it doesn't appear to me that networkd-dispatcher is built from a particular branch when running these tests. From what I can tell it's just whatever Ubuntu ships with. Unless @yuwata or others know better?

At any rate it seems like a chicken-and-egg problem here, although I must admit I haven't dove into the particular reason why networkd-dispatcher is failing.

Any thoughts on this? I'm not sure what more I can do besides debug networkd-dispatcher's failure and try to make a change to that project which allows it to start, but doesn't make any particular assumption about this new feature.

Thanks!

yuwata commented:
@sipraga and @craftguy Sorry for late response.

So, let's first implement `Describe()` DBus method in networkd and implement `networkctl status --json`, like ac8a4f6b78fe70680d3bc1704b1f85bbcbe7bb55. Then, networkd-dispatcher or any other tools which read the output of `networkctl` can easily and safely parse the output, and we can freely add any arbitrary columns to the output. WDYT?

I will implement `Describe()` and `networkctl status --json` later. Please wait for a while.

sipraga commented:
Sounds good @yuwata, please ping me once you have implemented `Describe()` and `networkctl status --json`. I'll then rebase and update the PR accordingly.

poettering commented on src/network/networkd-state-file.c:
```diff
@@ -445,9 +468,13 @@ int link_save(Link *link) {
 
+                online_state = link_online_state_to_string(link->online_state);
+                if (online_state)
+                        fprintf(f, "ONLINE_STATE=%s\n", strempty(online_state));
```
link_online_state_to_string() will return NULL if the state isn#t known/invalid, and a good non-empty string otherwise. hence you can drop the strempty() here

sipraga replied:
Thanks, I'll address this in the next revision (pending some other changes, see above)

yuwata commented:
@sipraga and @craftyguy I created a PR which implements `Describe()` bus method in networkd and `--json` option for networkctl. PTAL.

sipraga commented:
> @sipraga and @craftyguy I created a PR which implements `Describe()` bus method in networkd and `--json` option for networkctl. PTAL.

Looks good to me @yuwata. Once it's merged I'll rebase my changes here and add the online state to the `Describe()` JSON for manager and links.

Looking ahead, do you intend for us to wait for @craftyguy to update networkd-dispatcher and for this to make it into the Ubuntu CI before we can merge this?

craftyguy commented:
> @sipraga and @craftyguy I created a PR which implements `Describe()` bus method in networkd and `--json` option for networkctl. PTAL.

Nice!

Hmm, I'm not sure if I can implement the changes in networkd-dispatcher until [your patch](https://github.com/systemd/systemd/pull/19592) has landed in systemd and a release is available with it. Switching networkd-dispatcher to use `networkctl --json` will break compatibility with all previous systemd versions that don't have that option from your patch.

yuwata commented:
@craftyguy I guess the following procedure works for both newer and older `networkctl`:
1. first try to call `networkctl list --json=`. If succeeds, then parse the output and use it.
2. If that does not work, then you can assume that `networkctl` is old. So, try to call `networkctl list` and parse the text as already implemented.

WDYT?

yuwata commented on src/network/networkd-link.c:
```diff
@@ -256,6 +257,38 @@ void link_update_operstate(Link *link, bool also_update_master) {
                 operstate = LINK_OPERSTATE_ENSLAVED;
 
+        /* Only determine online state for managed links with RequiredForOnline=yes */
+        if (!link->network || link->network->unmanaged || !link->network->required_for_online)
```
Sorry, this is my fault. I previously said that you also need to check `unmanaged` flag, but it is not necessary. As, you can see e.g. `link_initialized_and_synced()`, if `network->unmanaged` is set, we do not assign the `network` object to `link`. So,
```
if (!link->network || !link->network->required_for_online)
```
is enough. Sorry.

craftyguy commented:
> @craftyguy I guess the following procedure works for both newer and older `networkctl`:
> 
> 1. first try to call `networkctl list --json=`. If succeeds, then parse the output and use it.
> 2. If that does not work, then you can assume that `networkctl` is old. So, try to call `networkctl list` and parse the text as already implemented.
> 
> WDYT?

Yeah I was kinda hoping to avoid carrying around both mechanisms, but what you propose is probably what I'll end up doing.

Is it possible to build/run only the new networkctl for testing purposes, without building all of systemd? (this question is probably better suited for elsewhere than here, if there's an IRC channel or something I can jump on there to ask)

sipraga commented:
@yuwata please take a look at the latest revision. Changes:
- rebased
- remove unnecessary check for `link->network->unmanaged`
- remove unnecessary `strempty()` call
- include `OnlineState` in networkctl JSON

In case the online state is invalid/unknown, the JSON value will be null, e.g.:
```
[root@image ~]# networkctl --json=pretty
{
	"Interfaces" : [
		{
			"Index" : 1,
			"Name" : "lo",
			"AlternativeNames" : [],
			"Type" : "loopback",
			"Driver" : null,
			"SetupState" : "unmanaged",
			"OperationalState" : "carrier",
			"CarrierState" : "carrier",
			"AddressState" : "off",
			"OnlineState" : null,
			"IPv4AddressState" : "off",
			"IPv6AddressState" : "off"
		},
		{
			"Index" : 2,
			"Name" : "host0",
			"AlternativeNames" : [],
			"Type" : "ether",
			"Driver" : "veth",
			"SetupState" : "configured",
			"OperationalState" : "routable",
			"CarrierState" : "carrier",
			"AddressState" : "routable",
			"OnlineState" : "online",
			"IPv4AddressState" : "routable",
			"IPv6AddressState" : "degraded",
			"NetworkFile" : "/usr/lib/systemd/network/80-container-host0.network"
		}
	]
}
```

I figured this is the correct way to do it, but let me know if you prefer something else like stringifying it to "unknown". Issues like this are why I initially added an explicit unknown state to the `LinkOnlineState` enum.

sipraga commented:
@craftyguy you can checkout systemd and build it locally and then just use the networkctl binary in `build/` on your running host. That way you can simulate a newer networkctl on an older networkd. Although I wouldn't expect to see such a combination in the wild, since the two are shipped together in the same package.

You can also use the mkosi VM which docs/HACKING.md describes. I had some trouble getting it to work on my Arch system, but it turned out I needed mkosi version 9. If you see some cryptic error like:
```
mkosi: error: argument verb: invalid choice: 'yes' (choose from 'build', 'clean', 'help', 'summary', 'shell', 'boot', 'qemu')
```
then you have to upgrade mkosi.

Then you can test against a locally build networkd/networkctl too.

yuwata commented on src/network/networkctl.c:
```diff
@@ -1612,6 +1639,13 @@ static int link_status_one(
                                    on_color_setup, strna(setup_state), off_color_setup);
         if (r < 0)
                 return table_log_add_error(r);
+        r = table_add_many(table,
```
Not a matter, but I prefer to add a blank line before `table_add_many()`.

yuwata commented on src/network/networkd-json.c:
```diff
@@ -59,6 +59,7 @@ int link_build_json(Link *link, JsonVariant **ret) {
                                         JSON_BUILD_PAIR("OperationalState", JSON_BUILD_STRING(link_operstate_to_string(link->operstate))),
                                         JSON_BUILD_PAIR("CarrierState", JSON_BUILD_STRING(link_carrier_state_to_string(link->carrier_state))),
                                         JSON_BUILD_PAIR("AddressState", JSON_BUILD_STRING(link_address_state_to_string(link->address_state))),
+                                        JSON_BUILD_PAIR("OnlineState", JSON_BUILD_STRING(link_online_state_to_string(link->online_state))),
```
Not a matter, but I prefer this is after `IPv6AddressState`.

yuwata commented on src/network/networkctl.c:
```diff
@@ -840,7 +862,9 @@ static int list_links(int argc, char *argv[], void *userdata) {
                                    TABLE_SET_COLOR, on_color_operational,
                                    TABLE_STRING, setup_state,
-                                   TABLE_SET_COLOR, on_color_setup);
+                                   TABLE_SET_COLOR, on_color_setup,
+                                   TABLE_STRING, online_state ?: "unknown",
+                                   TABLE_SET_COLOR, on_color_online);
```
Let's add the new 'online' column in later PR after @craftyguy implemented `--json` support in `networkd-dispatcher`.
Adding `Online` field in `status` command is OK.

yuwata reviewed (changes requested):
Looks good for me. But let's add the `online` column in `networkctl list` in later PR.

yuwata commented:
> > @craftyguy I guess the following procedure works for both newer and older `networkctl`:
> > 
> > 1. first try to call `networkctl list --json=`. If succeeds, then parse the output and use it.
> > 2. If that does not work, then you can assume that `networkctl` is old. So, try to call `networkctl list` and parse the text as already implemented.
> > 
> > WDYT?
> 
> Yeah I was kinda hoping to avoid carrying around both mechanisms, but what you propose is probably what I'll end up doing.
> 
> Is it possible to build/run only the new networkctl for testing purposes, without building all of systemd? (this question is probably better suited for elsewhere than here, if there's an IRC channel or something I can jump on there to ask)

@craftyguy Without container, I usually do something like the following to test networkd and networkctl from current git HEAD.
```
$ mkdir ~/git
$ cd ~/git
$ git clone https://github.com/systemd/systemd
$ cd systemd
$ git remote add sipraga git@github.com:sipraga/systemd.git
$ git fetch sipraga online-if-required
$ git checkout online-if-required
$ sudo mkdir /var/build/systemd
$ sudo chown USER:GROUP /var/build/systemd
$ cd /var/build/systemd
$ meson ~/git/systemd .
$ ninja
$ sudo systemctl edit systemd-networkd.service
$ cat /etc/systemd/system/systemd-networkd.service.d/override.conf
[Service]
ExecStart=
ExecStart=/var/build/systemd/systemd-networkd
$ sudo systemctl restart systemd-networkd.service
$ /var/build/systemd/networkctl status --json
```
Of course, you can also use mkosi or something similar to create a container or vm instead.

sipraga commented:
Thanks @yuwata for your patience with this PR. I'll split the networkctl patch then and resubmit the `ONLINE` column as a separate PR that we can merge when networkd-dispatcher is ready. This PR will still fix #19143, which was my original goal, so fine by me. I'll address your other comments too.

yuwata reviewed (approved):
LGTM. Thank you!

sipraga commented:
- rebased
- fix formatting & style comments
- remove online state column in `networkctl list` and reorder/refactor commits accordingly

Fingers crossed the Ubuntu CI passes this time.

I have a separate branch for `networkctl list` which I will make into a PR once this is merged: https://github.com/sipraga/systemd/compare/online-if-required...sipraga:networkctl-list-online

yuwata commented:
Rebased and force-pushed.
